File: ios/Runner/NativeGoogleMap.swift
import Flutter
import UIKit
import GoogleMaps

class NativeGoogleMap: NSObject, FlutterPlatformView, GMSMapViewDelegate {
  private let mapView: GMSMapView
  private let channel: FlutterMethodChannel

  init(frame: CGRect, viewId: Int64, args: [String: Any]?, messenger: FlutterBinaryMessenger) {
    let lat = args?["lat"] as? Double ?? 0
    let lng = args?["lng"] as? Double ?? 0
    let zoom = args?["zoom"] as? Double ?? 14
    let camera = GMSCameraPosition(latitude: lat, longitude: lng, zoom: Float(zoom))
    mapView = GMSMapView(frame: frame, camera: camera)
    channel = FlutterMethodChannel(name: "native_google_map_\(viewId)" , binaryMessenger: messenger)
    super.init()
    channel.setMethodCallHandler(handle)
    mapView.delegate = self
    channel.invokeMethod("mapReady", arguments: nil)
  }

  func view() -> UIView {
    return mapView
  }

  private func handle(_ call: FlutterMethodCall, result: FlutterResult) {
    switch call.method {
    case "moveCamera":
      if let args = call.arguments as? [String: Any],
         let lat = args["lat"] as? Double,
         let lng = args["lng"] as? Double,
         let zoom = args["zoom"] as? Double {
        let camera = GMSCameraPosition(latitude: lat, longitude: lng, zoom: Float(zoom))
        mapView.moveCamera(GMSCameraUpdate.setCamera(camera))
      }
      result(nil)
    case "setMarkers":
      mapView.clear()
      if let args = call.arguments as? [String: Any],
         let markers = args["markers"] as? [[String: Any]] {
        for m in markers {
          if let lat = m["lat"] as? Double, let lng = m["lng"] as? Double {
            let marker = GMSMarker(position: CLLocationCoordinate2D(latitude: lat, longitude: lng))
            if let rotation = m["rotation"] as? Double {
              marker.rotation = rotation
            }
            if let bytes = m["icon"] as? FlutterStandardTypedData {
              marker.icon = UIImage(data: bytes.data)
              marker.groundAnchor = CGPoint(x: 0.5, y: 0.5)
            }
            marker.map = mapView
          }
        }
      }
      result(nil)
    case "setPolylines":
      if let args = call.arguments as? [String: Any],
         let pts = args["polyline"] as? [[Double]] {
        let path = GMSMutablePath()
        for p in pts {
          if p.count == 2 {
            path.add(CLLocationCoordinate2D(latitude: p[0], longitude: p[1]))
          }
        }
        let poly = GMSPolyline(path: path)
        let color = (args["color"] as? NSNumber)?.intValue ?? 0xff4285F4
        poly.strokeColor = UIColor(rgb: UInt(color))
        poly.strokeWidth = CGFloat(args["width"] as? Double ?? 5.0)
        poly.map = mapView
      }
      result(nil)
    case "setPolygons":
      if let args = call.arguments as? [String: Any],
         let polys = args["polygons"] as? [[[Double]]] {
        for poly in polys {
          let path = GMSMutablePath()
          for p in poly { if p.count == 2 { path.add(CLLocationCoordinate2D(latitude: p[0], longitude: p[1])) } }
          let gPoly = GMSPolygon(path: path)
          let strokeColor = (args["strokeColor"] as? NSNumber)?.intValue ?? 0xff4285F4
          let fillColor = (args["fillColor"] as? NSNumber)?.intValue ?? 0x554285F4
          gPoly.strokeColor = UIColor(rgb: UInt(strokeColor))
          gPoly.fillColor = UIColor(rgb: UInt(fillColor))
          gPoly.strokeWidth = CGFloat(args["strokeWidth"] as? Double ?? 1.0)
          gPoly.map = mapView
        }
      }
      result(nil)
    case "setMapStyle":
      if let args = call.arguments as? [String: Any],
         let style = args["style"] as? String {
        _ = mapView.mapStyle(withJsonString: style)
      }
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - GMSMapViewDelegate
  func mapView(_ mapView: GMSMapView, didTapAt coordinate: CLLocationCoordinate2D) {
    channel.invokeMethod("onTap", arguments: ["lat": coordinate.latitude, "lng": coordinate.longitude])
  }

  func mapView(_ mapView: GMSMapView, didLongPressAt coordinate: CLLocationCoordinate2D) {
    channel.invokeMethod("onLongPress", arguments: ["lat": coordinate.latitude, "lng": coordinate.longitude])
  }
}

private extension UIColor {
  convenience init(rgb: UInt) {
    self.init(
      red: CGFloat((rgb >> 16) & 0xFF) / 255.0,
      green: CGFloat((rgb >> 8) & 0xFF) / 255.0,
      blue: CGFloat(rgb & 0xFF) / 255.0,
      alpha: 1.0)
  }
}

File: android/app/src/main/kotlin/com/example/my_project/NativeGoogleMap.kt
package com.quicky.ridebahamas

import android.content.Context
import android.graphics.BitmapFactory
import android.graphics.Bitmap
import android.view.View
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.MapView
import com.google.android.gms.maps.MapsInitializer
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import com.google.android.gms.maps.model.PolylineOptions
import com.google.android.gms.maps.model.PolygonOptions
import com.google.android.gms.maps.model.MapStyleOptions
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

class NativeGoogleMap(
    context: Context,
    messenger: BinaryMessenger,
    id: Int,
    params: Map<String, Any>?
) : PlatformView, MethodChannel.MethodCallHandler {

    private val mapView: MapView = MapView(context)
    private var googleMap: GoogleMap? = null
    private val channel = MethodChannel(messenger, "native_google_map_" + id)

    init {
        channel.setMethodCallHandler(this)
        mapView.onCreate(null)
        // Ensure the MapView lifecycle is properly started so the
        // GoogleMap instance can render correctly.
        // Some devices require onStart to be invoked before onResume.
        mapView.onStart()
        mapView.onResume()
        mapView.getMapAsync { g ->
            googleMap = g
            val lat = (params?.get("lat") as? Double) ?: 0.0
            val lng = (params?.get("lng") as? Double) ?: 0.0
            val zoom = (params?.get("zoom") as? Double) ?: 14.0
            g.moveCamera(CameraUpdateFactory.newLatLngZoom(LatLng(lat, lng), zoom.toFloat()))
            g.setOnMapClickListener { ll ->
                channel.invokeMethod("onTap", mapOf("lat" to ll.latitude, "lng" to ll.longitude))
            }
            g.setOnMapLongClickListener { ll ->
                channel.invokeMethod("onLongPress", mapOf("lat" to ll.latitude, "lng" to ll.longitude))
            }
            channel.invokeMethod("mapReady", null)
        }
        MapsInitializer.initialize(context)
    }

    override fun getView(): View = mapView

    override fun dispose() {
        // Mirror the lifecycle methods called in init to avoid
        // memory leaks and ensure the map cleans up correctly.
        mapView.onPause()
        mapView.onStop()
        mapView.onDestroy()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "moveCamera" -> {
                val lat = call.argument<Double>("lat") ?: 0.0
                val lng = call.argument<Double>("lng") ?: 0.0
                val zoom = call.argument<Double>("zoom") ?: 14.0
                googleMap?.moveCamera(CameraUpdateFactory.newLatLngZoom(LatLng(lat, lng), zoom.toFloat()))
                result.success(null)
            }
            "setMarkers" -> {
                val markers = call.argument<List<Map<String, Any>>>("markers")
                googleMap?.clear()
                markers?.forEach { m ->
                    val lat = (m["lat"] as Number).toDouble()
                    val lng = (m["lng"] as Number).toDouble()
                    val rotation = (m["rotation"] as Number?)?.toFloat() ?: 0f
                    val bytes = m["icon"] as? ByteArray
                    val opts = MarkerOptions().position(LatLng(lat, lng)).rotation(rotation).anchor(0.5f, 0.5f).flat(true)
                    bytes?.let {
                        val bmp = BitmapFactory.decodeByteArray(it, 0, it.size)
                        opts.icon(BitmapDescriptorFactory.fromBitmap(bmp))
                    }
                    googleMap?.addMarker(opts)
                }
                result.success(null)
            }
            "setPolylines" -> {
                val pts = call.argument<List<List<Double>>>("polyline")
                val color = call.argument<Int>("color") ?: 0xff4285F4.toInt()
                val width = (call.argument<Double>("width") ?: 5.0).toFloat()
                val opts = PolylineOptions().color(color).width(width)
                pts?.forEach { p -> opts.add(LatLng(p[0], p[1])) }
                googleMap?.addPolyline(opts)
                result.success(null)
            }
            "setPolygons" -> {
                val polys = call.argument<List<List<List<Double>>>>("polygons")
                val strokeColor = call.argument<Int>("strokeColor") ?: 0xff4285F4.toInt()
                val fillColor = call.argument<Int>("fillColor") ?: 0x554285F4
                val strokeWidth = (call.argument<Double>("strokeWidth") ?: 1.0).toFloat()
                polys?.forEach { poly ->
                    val opts = PolygonOptions().strokeColor(strokeColor).fillColor(fillColor).strokeWidth(strokeWidth)
                    poly.forEach { p -> opts.add(LatLng(p[0], p[1])) }
                    googleMap?.addPolygon(opts)
                }
                result.success(null)
            }
            "setMapStyle" -> {
                val style = call.argument<String>("style")
                googleMap?.setMapStyle(style?.let { MapStyleOptions(it) })
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }
}
